File: passenger/src/test/java/com/modsen/passenger/integration/BaseIntegrationTest.java
package com.modsen.passenger.integration;


import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Testcontainers;
import org.testcontainers.utility.DockerImageName;

@Testcontainers
public class BaseIntegrationTest {

    public static final String POSTGRES_IMAGE_NAME = "postgres:15-alpine";

    protected static final PostgreSQLContainer<?> postgresContainer;

    static {
        postgresContainer = new PostgreSQLContainer<>(DockerImageName.parse(POSTGRES_IMAGE_NAME));
        postgresContainer.start();
    }

    @DynamicPropertySource
    static void init(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", postgresContainer::getJdbcUrl);
        registry.add("spring.datasource.password", postgresContainer::getPassword);
        registry.add("spring.datasource.username", postgresContainer::getUsername);
    }
}

File: kt-passenger/src/test/kotlin/com/modsen/ktpassenger/integration/BaseIntegrationTest.kt
package com.modsen.ktpassenger.integration

import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Testcontainers
import org.testcontainers.utility.DockerImageName


@ActiveProfiles("test")
@Testcontainers
open class BaseIntegrationTest {

    companion object {
        private const val POSTGRES_IMAGE_NAME = "postgres:15-alpine"

        private val postgresContainer = PostgreSQLContainer(
            DockerImageName.parse(
                POSTGRES_IMAGE_NAME
            )
        )

        @JvmStatic
        @DynamicPropertySource
        fun init(registry: DynamicPropertyRegistry) {
            postgresContainer.start()
            registry.add("spring.datasource.url") { postgresContainer.jdbcUrl }
            registry.add("spring.datasource.password") { postgresContainer.password }
            registry.add("spring.datasource.username") { postgresContainer.username }
        }
    }
}
